src/objective/objective_gpu.cu

```diff
@@ -0,0 +1,222 @@
+// GPU implementation of objective function.  Necessary as evaluating the objective
+// function takes significant time when training on the GPU.
+
+#include <dmlc/omp.h>
+#include <xgboost/logging.h>
+#include <xgboost/objective.h>
+#include <cmath>
+#include <vector>
+
+#include "../common/device_helpers.cuh"
+#include "./loss.h"
+
+using namespace dh;
+
+namespace xgboost {
+namespace obj {
+
+
+DMLC_REGISTRY_FILE_TAG(objective_gpu);
+
+struct GPURegLossParam : public dmlc::Parameter<GPURegLossParam> {
+  float scale_pos_weight;
+  // declare parameters
+  DMLC_DECLARE_PARAMETER(GPURegLossParam) {
+    DMLC_DECLARE_FIELD(scale_pos_weight).set_default(1.0f).set_lower_bound(0.0f)
+        .describe("Scale the weight of positive examples by this factor");
+  }
+};
+
+// GPU kernel for gradient computation
+template<typename Loss>
+__global__ void get_gradient_k
+(bst_gpair *__restrict__ out_gpair,  bool *__restrict__ label_correct,
+ const float * __restrict__ preds, const float * __restrict__ labels,
+ const float * __restrict__ weights, int n, float scale_pos_weight) {
+  int i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i >= n)
+    return;
+  float p = Loss::PredTransform(preds[i]);
+  float w = weights == nullptr ? 1.0f : weights[i];
+  float label = labels[i];
+  if (label == 1.0f)
+    w *= scale_pos_weight;
+  if (!Loss::CheckLabel(label))
+    *label_correct = false;
+  out_gpair[i] = bst_gpair
+    (Loss::FirstOrderGradient(p, label) * w, Loss::SecondOrderGradient(p, label) * w);
+}
+
+// GPU kernel for predicate transformation
+template<typename Loss>
+__global__ void pred_transform_k(float * __restrict__ preds, int n) {
+  int i = threadIdx.x + blockIdx.x * blockDim.x;
+  if (i >= n)
+    return;
+  preds[i] = Loss::PredTransform(preds[i]);
+}
+
+// regression loss function for evaluation on GPU (eventually)
+template<typename Loss>
+class GPURegLossObj : public ObjFunction {
+  protected:
+  
+  float *preds_d_, *labels_d_, *weights_d_;
+  bool *label_correct_d_;
+  bst_gpair *out_gpair_d_;
+  int max_n_;
+  std::vector<cudaStream_t> streams_;
+  bool copied_;
+
+  // allocate device data for n elements
```

Comment:
Consider using the dvec and bulk allocator classes in common/device_helpers.cuh to manage device memory.